paper-dialog.html

```diff
@@ -70,40 +61,206 @@ <h2>Header</h2>
-    this.cancelAnimation();
-    this.playAnimation('entry');
-  },
-
-  _renderClosed: function() {
-    this.cancelAnimation();
-    this.playAnimation('exit');
-  },
-
-  _onNeonAnimationFinish: function() {
-    if (this.opened) {
-      this._finishRenderOpened();
-    } else {
-      this._finishRenderClosed();
+  Polymer({
+    is: 'paper-dialog',
+
+    behaviors: [
+      Polymer.PaperDialogBehavior
+    ],
+
+    listeners: {
+      'webkitAnimationEnd': '_onAnimationEnd',
+    },
+
+    properties: {
+      /**
+       * Set Entry Animation 
+       */
+      entryAnimation: {
+        value: "fade-in-animation",
+        type: String
+      },
+      /**
+       * Set Exit Animation
+       */
+      exitAnimation: {
+        type: String
+      },
+      _showing: {
+        type: Boolean,
+        value: false
+      }
+    },
+
+    /**
+     * @return {void}
+     */
+    attached: function () {
+      this._addListeners();
+    },
+
+    /**
+     * @return {void}
+     */
+    detached: function () {
+      this._removeListeners();
+    },
+
+    _renderOpened: function () {
+      this._showing = true;
+      this.cancelAnimation();
+      this._entryAnimation();
+    },
+
+    _renderClosed: function () {
+      this._showing = false;
+      this.cancelAnimation();
+      this._exitAnimation();
+    },
+
+    _addListeners: function () {
+      this.listen(this, 'animationend', '_onAnimationEnd');
+    },
+    
+    _removeListeners: function () {
+      this.unlisten(this, 'animationend', '_onAnimationEnd');
+    },
+
+    cancelAnimation: function () {
+      // Short-cut and cancel all animations and hide
+      this.classList.remove(this.entryAnimation);
+      this.classList.remove(this.exitAnimation);
+    },
+
+    _entryAnimation: function () {
+      if (!this.entryAnimation) {
+        this._onAnimationEnd();
+      } else {
+        this.classList.add(this.entryAnimation);
+        // If not animation _onAnimationEnd will not be called and will fail to change state
+        if (this._isSetAnimationValid()) {
+          this._onAnimationEnd();
+        }
+
+      }
+    },
+
+    _exitAnimation: function () {
+      if (!this.exitAnimation) {
+        this._onAnimationEnd();
+      } else {
+        this.classList.add(this.exitAnimation);
+        // If not animation _onAnimationEnd will not be called and will fail to change state
+        if (this._isSetAnimationValid()) {
+          this._onAnimationEnd();
+        }
+      }
+    },
+
+    _onAnimationEnd: function () {
+      // If no longer showing add class hidden to completely hide dialog
+      if (!this._showing) {
+        this.classList.remove(this.exitAnimation);
+        this._finishRenderClosed();
+      } else {
+        this._finishRenderOpened();
+      }
+    },
+
+    _isSetAnimationValid: function () {
+      // If CSS Class does not have animation-name considered not be an animation
+      var style = window.getComputedStyle(this);
```

Comment:
It would be worth to check for `animation-duration` in here too, e.g. if it is `0` then we don't have to wait for `animationend`. WDYT?